### PR TITLE
syncer: add unit test

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
### What problem does this PR solve? 
improve unit test coverage

### What is changed and how it works?
add unit test, coverage 24%->50%

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: syncer/db_test.go [hunk 1]
```diff
@@ -32,3 +33,18 @@ func (s *testSyncerSuite) TestGetServerID(c *C) {
...
+	c.Assert(files, Not(HasLen), 0)
+
+	pos := mysql.Position{
+		Name: files[len(files)-1].name,
+		Pos:  0,
+	}
+
+	remainingSize, err := countBinaryLogsSize(pos, s.db)
+	c.Assert(err, IsNil)
+	c.Assert(remainingSize, Equals, files[len(files)-1].size)
```
Comment (csuzhangxc) [inline syncer/db_test.go]: execute [FLUSH BINARY LOGS](https://dev.mysql.com/doc/refman/8.0/en/flush.html#flush-binary-logs) and to verify `getBinaryLogs`/`countBinaryLogsSize` again?

File: syncer/ddl_exec_info_test.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	c.Assert(ddlExecInfo.BlockingDDLs(), DeepEquals, ddls)
+
+	ddlExecInfo.ClearBlockingDDL()
+	c.Assert(ddlExecInfo.BlockingDDLs(), IsNil)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	err := ddlExecInfo.Send(ctx, newDDLExecItem(new(pb.ExecDDLRequest)))
+	c.Assert(err, IsNil)
```
Comment (csuzhangxc) [inline syncer/ddl_exec_info_test.go]: check status after `Send`?

File: syncer/ddl_exec_info_test.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+	ddlExecInfo.Close()
+	c.Assert(ddlExecInfo.status.Get(), Equals, ddlExecClosed)
```
Comment (csuzhangxc) [inline syncer/ddl_exec_info_test.go]: try `Renew` again after `Close`

File: syncer/dml_test.go [hunk 4]
```diff
@@ -44,3 +44,63 @@ func (s *testSyncerSuite) TestCastUnsigned(c *C) {
 		c.Assert(obtained, Equals, cs.expected)
 	}
 }
+
+func (s *testSyncerSuite) TestGenColumnPlaceholders(c *C) {
+	placeholderStr := genColumnPlaceholders(3)
+	c.Assert(placeholderStr, Equals, "?,?,?")
```
Comment (csuzhangxc) [inline syncer/dml_test.go]: I think you can also test with **1** `?` here.

File: syncer/dml_test.go [hunk 5]
```diff
@@ -44,3 +44,63 @@ func (s *testSyncerSuite) TestCastUnsigned(c *C) {
...
+			name: "a",
+		}, {
+			name: "b",
+		}, {
+			name: "c",
+		},
+	}
+
+	columnList := genColumnList(columns)
+	c.Assert(columnList, Equals, "`a`,`b`,`c`")
```
Comment (csuzhangxc) [inline syncer/dml_test.go]: test with only **1** column too?

File: syncer/heartbeat_test.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	if user == "" {
+		user = "root"
+	}
+	pswd := os.Getenv("MYSQL_PSWD")
+	t.from = config.DBConfig{
+		Host:     host,
+		User:     user,
+		Password: pswd,
+		Port:     port,
+	}
```
Comment (csuzhangxc) [inline syncer/heartbeat_test.go]: can we extract these lines into a util function for test cases like `obtainMySQLTestConfig`?

File: syncer/heartbeat_test.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+func (t *testHeartbeatSuite) TestHeartbeat(c *C) {
+	heartbeat, err := GetHeartbeat(&HeartbeatConfig{
+		serverID:       123,
+		masterCfg:      t.from,
+		updateInterval: int64(1),
+		reportInterval: int64(1),
+	})
+	c.Assert(err, IsNil)
+	err = heartbeat.AddTask("heartbeat_test")
```
Comment (csuzhangxc) [inline syncer/heartbeat_test.go]: can we test with multiple tasks?

File: syncer/job_test.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	}
+
+	for _, testCase := range testCases {
+		c.Assert(testCase.job.String(), Equals, testCase.jobStr)
+	}
+}
+
+func (t *testJobSuite) TestQueueBucketName(c *C) {
+	name := queueBucketName(9)
+	c.Assert(name, Equals, "q_1")
```
Comment (csuzhangxc) [inline syncer/job_test.go]: test more edge cases too? like `0`, `8`...

File: syncer/syncer_test.go [hunk 9]
```diff
@@ -983,3 +1000,227 @@ func (s *testSyncerSuite) TestGeneratedColumn(c *C) {
...
+
+func (s *testSyncerSuite) TestCasuality(c *C) {
+	s.cfg.WorkerCount = 1
+	syncer := NewSyncer(s.cfg)
+	syncer.jobs = []chan *job{make(chan *job, 1)}
+
+	go func() {
+		job := <-syncer.jobs[0]
+		c.Assert(job.tp, Equals, flush)
+		syncer.jobWg.Done()
```
Comment (csuzhangxc) [inline syncer/syncer_test.go]: any synchronization	mechanism to ensure this goroutine executed?

File: syncer/db_test.go (hunk 1 above)
Comment (WangXiangUSTC) [reply syncer/db_test.go]: done

File: syncer/ddl_exec_info_test.go [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+// limitations under the License.
+
+package syncer
+
+import (
+	"context"
+	"time"
+
+	. "github.com/pingcap/check"
+	"github.com/pingcap/dm/dm/pb"
```
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply syncer/ddl_exec_info_test.go]: done

File: syncer/ddl_exec_info_test.go (hunk 2 above)
Comment (WangXiangUSTC) [reply syncer/ddl_exec_info_test.go]: done

File: syncer/ddl_exec_info_test.go (hunk 3 above)
Comment (WangXiangUSTC) [reply syncer/ddl_exec_info_test.go]: done

File: syncer/dml_test.go (hunk 4 above)
Comment (WangXiangUSTC) [reply syncer/dml_test.go]: done

File: syncer/dml_test.go (hunk 5 above)
Comment (WangXiangUSTC) [reply syncer/dml_test.go]: done

File: syncer/syncer_test.go (hunk 9 above)
Comment (WangXiangUSTC) [reply syncer/syncer_test.go]: what do you mean?

File: syncer/heartbeat_test.go (hunk 7 above)
Comment (WangXiangUSTC) [reply syncer/heartbeat_test.go]: done

File: syncer/heartbeat_test.go (hunk 6 above)
Comment (WangXiangUSTC) [reply syncer/heartbeat_test.go]: done

File: syncer/job_test.go (hunk 8 above)
Comment (WangXiangUSTC) [reply syncer/job_test.go]: done

**Comment by WangXiangUSTC** [comment]
@csuzhangxc PTAL again

File: syncer/db_test.go [hunk 11]
```diff
@@ -47,4 +47,19 @@ func (s *testSyncerSuite) TestBinaryLogs(c *C) {
 	remainingSize, err := countBinaryLogsSize(pos, s.db)
 	c.Assert(err, IsNil)
 	c.Assert(remainingSize, Equals, files[len(files)-1].size)
+
+	s.db.Exec("FLUSH BINARY LOGS")
+	files, err := getBinaryLogs(s.db)
+	c.Assert(err, IsNil)
+	c.Assert(files, Not(HasLen), 0)
```
Comment (csuzhangxc) [inline syncer/db_test.go]: check files count already _+1_?

File: syncer/ddl_exec_info_test.go (hunk 2 above)
Comment (csuzhangxc) [reply syncer/ddl_exec_info_test.go]: if we can check `ddlExecSending` when sending, it's better. but maybe it's fine now.

File: syncer/syncer_test.go (hunk 9 above)
Comment (csuzhangxc) [reply syncer/syncer_test.go]: I mean, this goroutine may not execute if the main goroutine exited, then we do not whether `c.Assert(job.tp, Equals, flush)` passed.

Comment (WangXiangUSTC) [reply syncer/syncer_test.go]: you are right, done

File: syncer/db_test.go (hunk 11 above)
Comment (WangXiangUSTC) [reply syncer/db_test.go]: done

File: syncer/db_test.go [hunk 12]
```diff
@@ -32,3 +33,34 @@ func (s *testSyncerSuite) TestGetServerID(c *C) {
...
+
+	pos := mysql.Position{
+		Name: files[len(files)-1].name,
+		Pos:  0,
+	}
+
+	remainingSize, err := countBinaryLogsSize(pos, s.db)
+	c.Assert(err, IsNil)
+	c.Assert(remainingSize, Equals, files[len(files)-1].size)
+	fileNum := len(files)
```
Comment (GregoryIan) [inline syncer/db_test.go]: why not putting it before L42 to save some code

File: syncer/ddl_exec_info_test.go [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+	ddls := []string{"create database test"}
+	go func() {
+		select {
+		case _ = <-ddlExecInfo.Chan(ddls):
+		}
+	}()
+
+	time.Sleep(100 * time.Millisecond)
+	c.Assert(ddlExecInfo.BlockingDDLs(), DeepEquals, ddls)
```
Comment (GregoryIan) [inline syncer/ddl_exec_info_test.go]: I suggest to add a wait function to wait some event happen and retry

**Comment by GregoryIan** [comment]
Rest LGTM

File: syncer/ddl_exec_info_test.go (hunk 13 above)
Comment (WangXiangUSTC) [reply syncer/ddl_exec_info_test.go]: done https://github.com/pingcap/dm/pull/150/commits/4e689c7a16e3a3118df05cb11ded5ac48c0a3935

File: syncer/db_test.go (hunk 12 above)
Comment (WangXiangUSTC) [reply syncer/db_test.go]: done https://github.com/pingcap/dm/pull/150/commits/4e689c7a16e3a3118df05cb11ded5ac48c0a3935

**Comment by WangXiangUSTC** [comment]
@GregoryIan PTAL again

**Comment by WangXiangUSTC** [comment]
@GregoryIan PTAL again
